aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -518,32 +518,35 @@ const SparseTensor& resize_as_sparse_(const SparseTensor& self, const SparseTens
   return self;
 }
 
-SparseTensor dense_to_sparse(const Tensor& self, c10::optional<c10::Layout> layout, OptionalIntArrayRef blocksize) {
+SparseTensor dense_to_sparse(const Tensor& self, c10::optional<c10::Layout> layout, OptionalIntArrayRef blocksize, c10::optional<int64_t> dense_dim_opt) {
   if (layout.has_value()) {
     if (blocksize.has_value() && !(*layout == kSparseBsr || *layout == kSparseBsc)) {
       AT_ERROR("to_sparse for ", self.layout(), " to ", *layout, " conversion does not use specified blocksize");
     }
     if (self.layout() == *layout) {
       return self;
     }
+    if (dense_dim_opt.has_value() && !(*layout == kSparseCsr || *layout == kSparseCsc || *layout == kSparseBsr || *layout == kSparseBsc)) {
+      AT_ERROR("to_sparse for ", self.layout(), " to ", *layout, " conversion does not support specifying number of dense dimensions");
+    }
     switch (*layout) {
     case kStrided:
       return self;
     case kSparse:
       return dense_to_sparse(self, self.dim());
```

Comment:
To support `to_sparse(layout=torch.sparse_coo, dense_dim=dense_dim)`, use
```suggestion
    case kSparse:
      return dense_to_sparse(self, self.dim() - dense_dim_opt.value_or(0));
```